analyzer/emerald/incoming.go

```diff
@@ -344,23 +344,31 @@ func extractRound(b *block.Block, txrs []*sdkClient.TransactionWithResults, logg
 					ERC20Transfer: func(fromEthAddr []byte, toEthAddr []byte, amountU256 []byte) error {
 						amount := &big.Int{}
 						amount.SetBytes(amountU256)
-						if !bytes.Equal(fromEthAddr, common.ZeroEthAddr) {
+						fromZero := bytes.Equal(fromEthAddr, common.ZeroEthAddr)
+						toZero := bytes.Equal(toEthAddr, common.ZeroEthAddr)
+						if !fromZero {
 							fromAddr, err2 := registerRelatedEthAddress(blockData.AddressPreimages, blockTransactionData.RelatedAccountAddresses, fromEthAddr)
 							if err2 != nil {
 								return fmt.Errorf("from: %w", err2)
 							}
 							registerTokenDecrease(blockData.TokenBalanceChanges, eventAddr, fromAddr, amount)
 						}
-						if !bytes.Equal(toEthAddr, common.ZeroEthAddr) {
+						if !toZero {
 							toAddr, err2 := registerRelatedEthAddress(blockData.AddressPreimages, blockTransactionData.RelatedAccountAddresses, toEthAddr)
 							if err2 != nil {
 								return fmt.Errorf("to: %w", err2)
 							}
 							registerTokenIncrease(blockData.TokenBalanceChanges, eventAddr, toAddr, amount)
 						}
-						blockData.PossibleTokens[eventAddr] = &modules.EVMPossibleToken{
-							PossibleTypes: []modules.EVMTokenType{modules.EVMTokenTypeERC20},
-							EthAddr:       event.Address,
+						if _, ok := blockData.PossibleTokens[eventAddr]; !ok {
+							blockData.PossibleTokens[eventAddr] = &modules.EVMPossibleToken{}
+						}
+						// Mark as mutated if transfer is between zero address
+						// and nonzero address (either direction) and nonzero
+						// amount. These will change the total supply as mint/
+						// burn.
+						if fromZero != toZero && amount.Cmp(&big.Int{}) != 0 {
+							blockData.PossibleTokens[eventAddr].Mutated = true
```

Comment:
it's done as separate branches so that different ones can trigger in different iterations of the outer loop. one iteration can find a non-mutating event and create a default EVMPossibleToken{} and a later iteration can find a mutating event and turn on the Mutated flag

Reply:
ooh tricksy hobbitses
:+1: